### PR TITLE
[RFC] Implement cosmiconfig for workspace configuration

### LINKED ISSUE #62: Configuration file support
I think it would be valuable to support for some type of file that contains prettier's config so that projects can check it in source control.

I can see 2 cases where this would be useful.
- Allow tools to know that the project is using prettier. I'd like to enable thing like format on save but I don't want to reformat everyfile I touch on a project that doesn't use prettier.
- Allow per project config to avoid the need to change options when changing projects or to have to create custom formating scripts.

Could be it's own file (something like `.editorconfig`) or maybe part of `package.json`.

### LINKED ISSUE #918: [RFC] Editor Integration
Editors are the main place where we want people to use prettier, in order for it to be a great experience it should have the following properties:

- Use the version of prettier that the project uses. If not, fall back to an arbitrary version.
- When saving, use the options specified by the project, otherwise fall back to an arbitrary configuration.
- When reading a file that is inside of a project using prettier, and if an option is enabled, reformat the code using the local user preferences.

## Proposal

Allow people adding their configuration options inside of prettier in `package.json`:

```js
{
  ...

  "prettier": {
    "bracketSpacing": false,
    "trailingComma": "all"
  }
}
```

## How to find a package version of prettier?

Given a file with a path `a/b/c/d/e.js`, we want to crawl up the tree up until we find a file `package.json`.

It should check in this order:

- `a/b/c/d/package.json`
- `a/b/c/package.json`
- `a/b/package.json`
- `a/package.json`
- `package.json`

If it doesn't find one, then the file isn't inside of a prettier repo.

If it finds one in `a/b/package.json`, then look at the file `a/b/node_modules/.bin/prettier`. If it finds one, then use this prettier binary.

## How to deal with caching?

This is a fair amount of io to be done for each invocation. If you want to add caching, you want to be careful, here are the things to watch out for:

- `a/b/package.json` is updated
- `a/b/node_modules/.bin/prettier` is updated
- `a/b/c/package.json` or `a/b/c/d/package.json` is created

## How to find the configuration associated to a package version of prettier?

Given that a package version of prettier is always in `a/b/node_modules/.bin/prettier` and the package configuration is always in `a/b/package.json`, we can inside of prettier check if `../../package.json` exists, and if it does read it and use the prettier key.

If prettier is being used as a CLI with `require('prettier')`, I'm not sure how to find it and if it's even a good idea to.

## How do you resolve conflicting rules?

I suggest that we merge rules one by one and the order they are loaded in/override each other is:
- First load `package.json` rules
- Then load CLI arguments which override the one above

For example,

```js
// package.json
{ "prettier": {"trailingComma": "all", "singleQuote": true} }
```

```js
prettier --trailing-comma es5 --parser flow
```

would result in

```js
{singleQuote: true, trailingComma: 'es5', parser: 'flow'}
```

## How to make it fast?

Because all the integrations now have to use the cli and can no longer just use the local require version, we're going to pay js/node startup time on every single invocation.

I recommend adding an option to prettier bin `--json-rpc`. With this, you can open the prettier process once and send formatting commands through stdin/stdout without having to restart it every time.

A sample connection looks like this:

```js
-->
Content Length: xxx
{"jsonrpc": "2.0", "method": "format", "params": ["abc", {"trailingComma": "es6"}], "id": 1}

<--
Content Length: xxx
{"jsonrpc": "2.0", "result": {"formatted": "abc;", "error": null}, "id": 1}
```

## Conclusion

Let me know what you think :)

### PR DESCRIPTION
**Pull request by azz**
This RFC PR uses [cosmiconfig](https://github.com/davidtheclark/cosmiconfig) to support configuration in the following formats:

* `.prettierrc` (in JSON or YAML format)
* `prettier.config.js` exporting an object
* `package.json`'s `"prettier"` key.

Configuration is resolved from the file being formatted, up the tree.

The order of precedence (lowest to highest) is:
* Prettier defaults
* Options from a config file
* CLI options

~Addresses most of~ Closes #918 ~(minus the JSON RPC)~
Similar to #1086
Closes #62

### CODE REVIEW

**Comment by vjeux** [comment]
Could you do some small performance tests to see what the overhead is?

**Comment by vjeux** [comment]
For json-rpc, I found another way to do it so it's not needed anymore as part of prettier.

**Comment by CiGit** [comment]
The question which came up some times was: 
Is it possible to have different settings per language? (In an editor integration at least)
If I remember correctly @azz asked it :-)

**Comment by rattrayalex** [comment]
I have to say, `.prettierrc` doesn't sound very good. Does `cosmiconfig` support different filenames (eg; `.prettier`)?

Supporting a dynamic `prettier.config.js` also seems like overkill for prettier, for which settings presumably do not change between commits (unlike, say, webpack). 

What are the advantages to this approach over https://github.com/prettier/prettier/pull/1086 ?

**Comment by vjeux** [comment]
In order to find the files, we should do the queries in parallel. From the current path /a/b/c/, we can search for `/a/b/c/.prettier`, `/a/b/.prettier`, `/a/.prettier` and `/.prettier` in parallel so that we don't need to do a serial search.

Then, for the node version, we should watch all those files so that we don't need to do all those stats for every single format.

**Comment by josephfrazier** [comment]
@CiGit:

> Is it possible to have different settings per language? (In an editor integration at least)

I don't think there's been a specification proposed for this, but we could always extend the configuration format later if needed. Until then, you could work around it by keeping files with different languages in their own directories, with corresponding configuration files (I realize this isn't great).

---

@rattrayalex:

> I have to say, `.prettierrc` doesn't sound very good.

Hmm, but `.*rc` is the convention for files like this (`.eslintrc`, `.npmrc`, `.yarnrc`, etc), so I think we should stick to it. It looks like `cosmiconfig` supports a `--config` flag, so you could pass a different filename in if you'd like.

---

@vjeux:

> Then, for the node version, we should watch all those files so that we don't need to do all those stats for every single format.

Wouldn't this require implementing a client/server architecture? I think a good first step is to simply make it possible to do one-off formatting that uses configuration files without having to address the client/server issue, as this PR or https://github.com/prettier/prettier/pull/1086 would do. If performance is a concern, we could hide the config-reading behind a flag, so that users without config files don't have to wait for the unnecessary IO.

> For json-rpc, I found another way to do it so it's not needed anymore as part of prettier.

I'm curious, what are you doing instead?

**Comment by vjeux** [comment]
I'm just copying this file in the repo :p So it doesn't have to be within prettier anymore.
https://github.com/prettier/prettier-rpc/blob/master/prettier-rpc.js

**Comment by josephfrazier** [comment]
Oh yeah, that project :) I remember trying to figure out how to use it, but I couldn't quite crack it. See also https://github.com/prettier/prettier-rpc/issues/1

**Comment by vjeux** [comment]
The biggest reason I have waited so long to get configuration is performance and correctness.

There are many "easy" ways to deal with it:
 - Always do the search
 - Do it only once at the beginning
 - Push it to all the integrations which can ask the user to reload when it changes
 - Add a "caching" layer

All of them kind of work but are super confusing when someone upgrades prettier/config to your repo. This is probably not a huge deal for prettier itself because you're going to do it very infrequently, but at Facebook, this kind of issues is the most annoying for developers and the hardest to solve for people working on implementations.

If we add a way to configure prettier, we need to do it right. If you don't upfront, then it's never going to be seen as high pri enough and never going to be done. Also, based on which approach you take, it may be harder to do later.

We need two properties:

- Performance: be close to a no-op the second time you call format()
- Correctness: if you add/remove/update the config file, during the next format() call, it must be taken into account.

**Comment by CiGit** [comment]
I should have given some use case our users (prettier-vscode) and by extension prettier's users. They mostly concerne trailing comma
1. - build files (webpack, gulp, ...) at the root, JS files no trailing comma (node < 8)
    - Typescript project in a *src* folder.
2. - Same as 1 but build files in a *script* folder.
3. - 1 or 2
    - Mixed source files between JS / Typescript (project being migrated to typescript)

@josephfrazier 2 would be covered by the example you gave. For 1, is there any *cascading* rule ? Those it take into account only the first encountered config file?

3 is the most problematic here.
I've an idea to implement it in vscode, but I would prefer to have those configurations handled by prettier itself to avoid having different implementations across editors which is a bad thing to me. I would expect all of them to work the same way without having to configure your editor first.

**Comment by josephfrazier** [comment]
@CiGit:

> For 1, is there any cascading rule ? Those it take into account only the first encountered config file?

Hmm, it looks like cosmiconfig *explicitly does not* merge configuration files together. See https://github.com/davidtheclark/cosmiconfig/tree/334065d9f7b557170a38b3f920bf5e90cecfab72#differences-from-rc:

> * Stops at the first configuration found, instead of finding all that can be found down the filetree and merging them automatically.

One way to fix this might be to switch from cosmiconfig to [config-attendant](https://github.com/GarthDB/config-attendant), which appears to merge configurations together: https://github.com/GarthDB/config-attendant/blob/f435d03b32aab568b336e1b47942bcfefea2fc0f/src/lib/index.js#L104-L110

> 3. Mixed source files between JS / Typescript (project being migrated to typescript)

Hmm, if you want to be able to specify different configurations for *.js and *.ts files in the same directory, I think you could use eslint with eslint-plugin-prettier, as it looks like eslint just recently merged support for glob-based configuration: https://github.com/eslint/eslint/commit/a21dd32c46f95bc232a67929c224824692f94b70

Otherwise, we'd have to come up with a spec for supporting this in package.json/.prettierrc/etc.

---

(forgot this earlier) @rattrayalex:

> What are the advantages to this approach over #1086 ?

For one, it supports configuration in files other than package.json. This has been suggested a few times:

* https://github.com/prettier/prettier/issues/918#issuecomment-298797387
* https://github.com/prettier/prettier/issues/918#issuecomment-289199162
* https://github.com/prettier/prettier/issues/918#issuecomment-294191768
* https://github.com/prettier/prettier/issues/918#issuecomment-313272241

**Comment by CiGit** [comment]
The merging thing was not that important I think. There are not that much options :-)
May be worth exploring eslint's glob config. Even if typescript and eslint work pretty badly together (At least last time I tried something with both)

**Comment by lydell** [comment]
> looks like eslint just recently merged support for glob-based configuration

Just as a quick aside, [Prettier even uses that glob-based configuration](https://github.com/prettier/prettier/blob/0fb332b2f95795484e35c050d83fade48d8a00c0/scripts/build/.eslintrc.js).

**Comment by rattrayalex** [comment]
I'm 👍 to not-merging, and just using the closest config. More explicit and I think would play better with large, nested repos. 

I made a propsal regarding perf in the wrong thread, which I will copy here....

**Comment by rattrayalex** [comment]
(copied from https://github.com/prettier/prettier/issues/918#issuecomment-313872837 )

Alright, so regarding perf of finding the config file...

I think there are two main uses of prettier (correct me if I'm wrong!): from the project root, eg in CI/precommit/manually/etc across a whole directory, or via an editor integration while writing. Obviously, the I/O overhead in the first case is probably-usually-fine, as long as it's done once per run instead of once per child file. The perf savings of finding the config file are only important (I take it) for the editor integrations.

In that case, for editor integrations, I propose we just add a --config-file option, which takes a filepath, and when it is provided, use it without checking for others. Editors can then be responsible for caching/providing it.

It would then presumably also be helpful to expose prettier find-config FILEPATH or similar, so that editors don't have to find the config files themselves (and thus keeping it consistent across editors).

I think this will address the perf concerns while keeping the code within prettier quite simple (eg; we can just use cosmiconfig or whatever).

Thoughts?

**Comment by josephfrazier** [comment]
@vjeux:

> Could you do some small performance tests to see what the overhead is?

Out of curiousity, I decided to go ahead and do this upon prettier's own codebase on both master (commit 0fb332b2f95795484e35c050d83fade48d8a00c0) and this PR (commit 53c3e38161d159d9de7e11d6ff1e4e03d3b5a13f) by running:

    perf stat -r 10 -d ./bin/prettier.js --write src/*

On master, I got this result:

       1.274921339 seconds time elapsed                                          ( +-  0.43% )

Whereas with this PR, I got the following:

       1.348827530 seconds time elapsed                                          ( +-  0.80% )

Which represents a roughly 5% increase in runtime over the 271655 bytes of code across 25 files in src/. This could probably be optimized by only loading the configuration once per directory, rather than for each file.

**Comment by josephfrazier** [comment]
@vjeux:

> Could you do some small performance tests to see what the overhead is?

Out of curiousity, I decided to go ahead and do this upon prettier's own codebase on both master (commit 0fb332b2f95795484e35c050d83fade48d8a00c0) and this PR (commit 53c3e38161d159d9de7e11d6ff1e4e03d3b5a13f) by running:

    perf stat -r 10 -d ./bin/prettier.js --write src/*

On master, I got this result:

       1.274921339 seconds time elapsed                                          ( +-  0.43% )

Whereas with this PR, I got the following:

       1.348827530 seconds time elapsed                                          ( +-  0.80% )

Which represents a roughly 5% increase in runtime over the 271655 bytes of code across 25 files in src/. This could probably be optimized by only loading the configuration once per directory, rather than for each file.

---

(Edited to add) @rattrayalex:

> I think there are two main uses of prettier (correct me if I'm wrong!): from the project root, eg in CI/precommit/manually/etc across a whole directory, or via an editor integration while writing. Obviously, the I/O overhead in the first case is probably-usually-fine, as long as it's done once per run instead of once per child file. The perf savings of finding the config file are only important (I take it) for the editor integrations.

That's a good point. For what it's worth, if you're already using eslint in conjunction with prettier, you can add [eslint_d](https://github.com/mantoni/eslint_d.js) into the mix to get configuration caching without having to wait on prettier to implement it. There's some more discussion starting here: https://github.com/prettier/prettier/issues/918#issuecomment-294017348

If you're not using eslint, or find eslint_d to be too slow, you could also try [prettier_d](https://github.com/josephfrazier/prettier_d), which is focused on prettier rather than eslint.

**Comment by rattrayalex** [comment]
@josephfrazier looks like `prettier_d` is yours (it looks great!). Would you be willing to:

1. Adopt whatever config-file-finding/parsing protocol is adopted here
2. Transfer the package to the prettier org

?

Not sure what @vjeux thinks of it.

**Comment by josephfrazier** [comment]
Thanks @rattrayalex! I've been meaning rewrite prettier_d as a fork of prettier, in order to make it easier to keep up with changes (new options, etc), as well as to fix the current [limitations](https://github.com/josephfrazier/prettier_d/tree/158fb689a3e8b5c9c577da6171234d9463c33926#limitations) of prettier_d. For reference, here are some relevant issues:

* https://github.com/prettier/prettier/pull/1352 Export minimist options
* https://github.com/prettier/prettier/pull/2168 [RFC] Expose CLI programmatically

This would make it easier to unify the configuration handling as well. If @vjeux, etc. are on board, I'd be willing to consider moving it to the prettier organization once the rewrite is done (or before, if that's ok). Note that prettier_d is already listed in the README under Related Projects (and I just edited it to mention the configuration support: https://github.com/prettier/prettier/commit/1e5ae50eba47142a02cc21e46864a9d79e29afed), so hopefully it's still somewhat discoverable.

**Comment by azz** [comment]
How do we feel about extending the configuration format to include `extname` support:

```yaml
trailingComma: es5

.ts:
  trailingComma: all
```

```yaml
{
  "trailingComma": "es5",
  ".ts": {
    "trailingComma": "all"
  }
}
```

And a method to the API:

#### `prettier.resolveOptions(file, { useCache }): Promise<PrettierOptions>`

Returns a promise resolving to an object to be passed to `prettier.format()`. 

* When `file` is omitted, uses the current working directory.
* `useCache` defaults to true.

cc. @CiGit

**Comment by robwise** [comment]
I'm slightly confused here. Are we saying prettier is going to automatically get its own options when I try to run it or is it going to have a `resolveOptions` method that lets me get them first, modify them if I choose, and then make another call to prettier, passing in the options, to actually format? 

The reason I ask is that when developing prettier-atom, I have often had to override certain settings in special cases. For example, sometimes users choose to have their line length or tabs-vs-spaces preference inferred from Atom's setting for the given scope they are in. If I can't interject these settings in the case that these rules cannot be inferred from a config, then the prettier-atom package is going to lose a lot of that functionality.

**Comment by azz** [comment]
@robwise  So the flow would be:

```js
let options = await prettier.resolveOptions(fileBeingSaved); // traverse filesystem to find .prettierrc, etc
options = applyPrettierAtomOverrides(options); // add special case options as you mentioned
const output = prettier.format(read(fileBeingSaved), options);
```

**Comment by azz** [comment]
@rattrayalex 

> I think there are two main uses of prettier (correct me if I'm wrong!): from the project root, eg in CI/precommit/manually/etc across a whole directory, or via an editor integration while writing. Obviously, the I/O overhead in the first case is probably-usually-fine, as long as it's done once per run instead of once per child file. The perf savings of finding the config file are only important (I take it) for the editor integrations.

It is not once-per-run, as we want to be able to run prettier on a glob:

```
packages/
  foo/
    .prettierrc
    index.js
  bar/
    .prettierrc
    index.js
```

`prettier --write 'package/**/*.js` and have that work as a expected.

> In that case, for editor integrations, I propose we just add a --config-file option, which takes a filepath, and when it is provided, use it without checking for others. Editors can then be responsible for caching/providing it.

Cosmiconfig supports a `--config` CLI option out of the box! :tada:

> It would then presumably also be helpful to expose prettier find-config FILEPATH or similar, so that editors don't have to find the config files themselves (and thus keeping it consistent across editors).

Done, as an API, see above, I can also add a CLI version: `prettier --resolve-config <filepath>`.

**Comment by robwise** [comment]
@azz Oh okay awesome, that sounds good to me!

**Comment by azz** [comment]
Added:

* `--resolve-config <path>` CLI option
  Prints the path to the resolved configuration file, relative to the current working directory. If no file is found, nothing is written and the exit code is `1.

* `--config <config-file>` CLI option
  Shortcuts the config resolution logic.

So an editor integration using the CLI can do:

```bash
# cache $config_file
config_file=$(prettier --resolve-config $input_file)
prettier --config "$config_file" --write "$input_file"
```

If using the API, https://github.com/prettier/prettier/pull/2434#issuecomment-313892495 applies instead.

**Comment by rattrayalex** [comment]
Cool! I think we should wait for @vjeux but this all looks good to me!

File: bin/prettier.js [hunk 1]
```diff
@@ -78,9 +89,57 @@ if (write && argv["debug-check"]) {
...
+  return resolver
+    .resolveConfig(filePath, { configFile: argv["config"] })
+    .then(options => {
+      const parsedArgs = minimist(args, {
+        boolean: booleanOptionNames,
+        string: stringOptionNames,
+        default: Object.assign(
+          {
+            semi: true,
+            color: true,
```
Comment (azz) [inline bin/prettier.js]: Note to self: `color` should be in the other list

**Comment by azz** [comment]
I still want to do a perf regression test against a project with a lot of files, maybe the babel monorepo.

**Comment by azz** [comment]
Tested the change on babel, on Windows will the following powershell command:

```powershell
Measure-Command {node ..\prettier\bin\prettier --write --trailing-comma es5 'packages/**/*.js'}
```

Type | Hash | Time
----|----|----
With cosmiconfig | d6bbdc6 | **9290ms**
Without cosmiconfig | 3728749 | **8103ms**

Around 15% slower. I tried passing `dirname()` to `cosmiconfig` instead but it didn't make a difference.

**Comment by CiGit** [comment]
@azz seems good to me.
What does `prettier.resolveConfig` return when there is no config file? When some options are missing from the file? We certainly want to fallback to editor's configuration at some point.

Additionally, is `.test.js` a valid extension? And what about `**/test/*.js`
Depending on the project's structure, this is something which can happen.

**Comment by azz** [comment]
> What does prettier.resolveConfig return when there is no config file? 

```
{}
```

> When some options are missing from the file? We certainly want to fallback to editor's configuration at some point.

Only options listed in the config file are returned.

> Additionally, is .test.js a valid extension? And what about **/test/*.js
Depending on the project's structure, this is something which can happen.

At the moment, the extension is just path.extname`, so `foo.test.js` is `.js`. We could consider using globs instead, if it is a common use case.

**Comment by CiGit** [comment]
That's perfect for a fallback @azz Thanks!

I don't know if it's common, but from [Jest: Getting Started](https://facebook.github.io/jest/docs/en/getting-started.html)
`sum.js`
`sum.test.js`

[Jest configuration](https://facebook.github.io/jest/docs/en/configuration.html#testregex-string)
> testRegex [string]
Default: `(/__tests__/.*|(\\.|/)(test|spec))\\.jsx?$`

**Comment by azz** [comment]
On second thoughts I'm changing the resolved value from `{}` to `null` if no configuration is found. This makes it possible to only apply editor-based settings if no config file is found.

**Comment by azz** [comment]
Just pushed another change, makes the overrides API almost identical to that of eslint.

```yaml
# .prettierrc
semi: false

overrides:
- files:
  - "*.test.js"
  - "**/__tests__/*.js"
  options:
    semi: true
```

I don't know that such fine-grained control is really required for prettier, but maybe it is better to have something flexible than breaking people if we go for something too simple and need to change it.

**Comment by robwise** [comment]
> On second thoughts I'm changing the resolved value from {} to null if no configuration is found. This makes it possible to only apply editor-based settings if no config file is found.

Yeah but if you do `{}` I can do:

```js
const prettierConfigOptions = await prettier.resolveConfig(fileBeingSaved);
const finalOptions = Object.assign({}, prettierAtomFallbackOptions, prettierConfigOptions);
const output = prettier.format(read(fileBeingSaved), options);
```

**Comment by ljharb** [comment]
@robwise you can do `Object.assign({}, null, undefined, false)` just fine.

**Comment by robwise** [comment]
@ljharb Oh roger that, thanks for the tip! Love your stuff btw

**Comment by azz** [comment]
How do we feel about the currently implemented eslint-style pattern overrides?

```json
{
  "printWidth": 80,
  "overrides": [
    {
      "files": "./some/minmatch/*/pattern.js",
      "excludeFiles": [ "not-this-pattern-.js", "or-this.js"],
      "options": {
        "printWidth": 100,
        "parser": "flow"
      }
    },
    {
      "files": ".prettierrc",
      "options": {
        "parser": "json"
      }
  ]
}
```

If we're generally happy with it we can merge this and keep iterating! :tada:

**Comment by robwise** [comment]
Seems fine to me.

One issue that came up with prettier-atom was what to do in the case that a config only specifies certain rules. Some people wanted to fall back to their editor's settings, because this implies that the rule is not enforced and up to the developer (or that the project hasn't defined a config yet), and others want to ignore the user's atom settings and fall back to the prettier defaults.

**Comment by azz** [comment]
@robwise 

> Some people wanted to fall back to their editor's settings, because this implies that the rule is not enforced and up to the developer (or that the project hasn't defined a config yet), and others want to ignore the user's atom settings and fall back to the prettier defaults.

That's what I like about this design, prettier will give the editor the API to resolve canonical overrides   from config files. As an editor integration you get to decide which order to apply these settings (before or after global/workspace settings, or in isolation).

`prettier-vscode` currently can resolve settings from a project's `settings.json` file, and may want to handle this differently to `prettier-atom`.

**Comment by azz** [comment]
Merged!

I'll write up some documentation tomorrow.

@CiGit @robwise if you want to give this a test run with your extensions on `prettier/prettier#master` and give some early feedback that would be amazing.

**Comment by CiGit** [comment]
@azz :+1: I will try to get a vsix test package under 50h

**Comment by CiGit** [comment]
Published [here](https://github.com/CiGit/prettier-vscode/releases/tag/v0.22.0-alpha)
Caching is not handled for now.

Doc is missing from https://github.com/prettier/prettier#api

**Comment by CiGit** [comment]
I'm trying to handle the cache and it seems I've misunderstood `{ useCache }`. (Now I've glanced at this code.)
This PR uses 2 instances of cosmiconfig, with/without cache.
Concerning `withCache`, we have no options to [clearCache](https://github.com/davidtheclark/cosmiconfig#clearcaches).
In a long running prettier, ie editor ..., if a file changes between 2 runs, cache will hold first version.
In my case, I have following options:
- Don't use cache `{ useCache: false}`
- Ask the user to, at least, reload vscode on config change.
- (handle cache myself)

I'd like to be able to call `prettier.wipeConfigCache()` or any other alternative to clear that config cache when a config file changes.

Is there a JSON Schema to validate those config files?

**Comment by azz** [comment]
Adding `clearConfigCache` in #2451, does that look good to you?

**Comment by azz** [comment]
I created a schema. @CiGit do you want to review it? https://github.com/SchemaStore/schemastore/pull/313

**Comment by CiGit** [comment]
Will take a look at it. I didn't know SchemaStore :-)
Would it be eventually bundled with prettier?
I could also provide a typescript .d.ts file if this is usefull.

**Comment by azz** [comment]
How would you usually go about consuming the schema?

**Comment by CiGit** [comment]
I thought I would provide it with the extension and validate those files. I've not explored this yet. It seems to be possible to provide an URL.

**Comment by azz** [comment]
Cool. URL will be http://json.schemastore.org/prettierrc if it is merged.

**Comment by gaui** [comment]
When will this be released? 👍

**Comment by danawoodman** [comment]
@vjeux is there a release scheduled that would include this change?

**Comment by vjeux** [comment]
@gaui @danawoodman Sorry about that. I switched team at Facebook and took a lot of vacation in the past month, so didn't have a lot of time to work on prettier. I just opened an issue to start tracking the work needed to get this out to the world: https://github.com/prettier/prettier/issues/2589

**Comment by danawoodman** [comment]
Thanks @vjeux!

**Pull request merged**